### DISCUSSION TITLE
Loops?

### DISCUSSION THREAD

**Discussion by NathanielCJ99** [sqlpage/SQLPage#306] [Q&A] [answered]
Want to implement a loop function so I can make a variable number of data entry points(typed by user) in a form. Does not seem as any loops are usable/available.

Using this to connect to a testing database, based on the number of values needed per test. Any suggestions?

**Comment by lovasoa** [comment]
Hello and welcome to SQLPage!


sql is a declarative language. it doesn't have loops. 

However, what you want to do is definitively possible to express in sql using a [recursive common table expression](https://learnsql.com/blog/sql-recursive-cte/)

```sql
select 'form' as component;

WITH RECURSIVE InputFields AS (
  SELECT 1 AS i
  UNION ALL
  SELECT i + 1
  FROM InputFields
  WHERE i < $number_of_fields::int
)
SELECT 'Input field ' || i AS name
FROM InputFields;
```

**Reply by NathanielCJ99** [reply]
Thanks for quick reply. Would this work with a Microsoft SQL server? I tried the above Query and get this error. "error returned from database: Incorrect syntax near 'InputFields'."

**Reply by lovasoa** [reply, marked as answer]
Microsoft SQL Server does indeed support recursive CTEs, but not the keyword "recursive".

**Reply by NathanielCJ99** [reply]
Wonderful. That should allow me to make some more progress. Thanks!

**Reply by lovasoa** [reply]
you'll find the full documentation on Microsoft's website: https://learn.microsoft.com/en-us/sql/t-sql/queries/with-common-table-expression-transact-sql?view=sql-server-ver16

**Comment by NathanielCJ99** [comment]
Adding to this, I am wondering how I could insert the data from My from into the database after submission. Using the same WITH statement?

**Reply by lovasoa** [reply]
You may want to take a look at this example from the documentation: https://github.com/lovasoa/SQLpage/tree/main/examples/forms%20with%20a%20variable%20number%20of%20fields

**Reply by NathanielCJ99** [reply]
That does not quite seem to cover what I am looking for. I am trying to format this insert statement to allow my variable 'i' to be used to insert my values "Test Step 1-5". I can get them all individually, but I cannot seem to get the formatting correct to allow the 'i' to be used.

The number of test steps may be different values  


![database code snippet](https://github.com/lovasoa/SQLpage/assets/63015583/50e7706a-3995-42cb-87ba-18d771f2cb0e)

**Reply by lovasoa** [reply]
The example does what you need here: 
https://github.com/lovasoa/SQLpage/blob/main/examples/forms%20with%20a%20variable%20number%20of%20fields/order_insert.sql

The example uses the SQLite database, so you will need to adapt it to Microsoft SQL server using https://learn.microsoft.com/en-us/sql/relational-databases/json/json-data-sql-server?view=sql-server-ver16

**Reply by lovasoa** [reply]
![image](https://github.com/lovasoa/SQLpage/assets/552629/984bfdac-02f5-4ec5-87c3-d1f700d39b23)


an alternative if you want something more cross-database

```sql
with nums(i) as (
   select 1
   union all select i+1 from nums
  where i < 2
)
insert into test(x)
select JSON_VALUE(
  '{"field_1": "hello", "field_2": "world"}',
  CONCAT('$.field_', i)
)
from nums;
```

and you can replace the hardcoded `'{"field_1": "hello", "field_2": "world"}'` with a call to [sqlpage.variables](https://sql.ophir.dev/functions.sql?function=variables#function)

**Reply by NathanielCJ99** [reply]
I think I have almost got it... I am confused with what SQL page calls the array of my form results.(To replace the hardcoded section) That seems to be the piece of the puzzle I am missing. To input as the first argument for JSON_VALUE()

For the example given for sqlpage.variables
![image](https://github.com/lovasoa/SQLpage/assets/63015583/98600f98-2d16-41a4-9b06-59bfbc495ea9)

**Reply by lovasoa** [reply]
A call to `sqlpage.variables()` will return the json object you need, with form field names as keys and the user input as values.

**Reply by NathanielCJ99** [reply]
I got it to work. The spaces in my name were giving me issues... I changed them to under scores and now it works!

![image](https://github.com/lovasoa/SQLpage/assets/63015583/d2f9ef27-df41-4cd4-becb-c5eaf5ec07e4)

**Reply by rhodelta66** [reply]
@lovasoa; @NathanielCJ99 So funny to see this; We are building web apps with 100% sql too, but we can only build solutions for Microsoft SQL Server. Maybe we can join forces somehow so we can both be billionaires haha ​​(I saw your video on youtube)

**Reply by rhodelta66** [reply]
This is an example of action code we use:

```sql
--declare the switch variable and sow it on the modal page
DECLARE @moon_earth_switch nvarchar(max)='false'; EXEC sp_api_modal_get_value @name='@moon_earth_switch',@value = @moon_earth_switch OUT 		
EXEC sp_api_modal_switch @name = '@moon_earth_switch', @value = @moon_earth_switch OUT,@label = N'Earth - Moon switch',@direct_post = 1

--show text on the page						
DECLARE @t nvarchar(max)=CASE WHEN @moon_earth_switch = 'true' THEN N'Hello Moon' ELSE N'Hello Earth' END
EXEC sp_api_modal_text @T ,@class='h2'						

--send text to toast (bottom right) and audible output (speakers)
EXEC sp_api_toast @text=@t, @speech=@t, @speech_lang=N'en-en' --exit button (action ends)

--add a quit button
DECLARE @exit nvarchar(max); EXEC sp_api_modal_get_value @name='@exit'	,@value = @exit OUT 
EXEC sp_api_modal_button @name='@exit',@value = 'Exit', @valueout = @exit OUT,@class='btn-danger',@key='Esc'
IF @exit is not NULL EXEC sp_api_modal_clear
```
